zerver/actions/create_user.py

```diff
@@ -209,24 +228,23 @@ def add_new_user_history(user_profile: UserProfile, streams: Iterable[Stream]) -
         # Exclude the already-used ids and sort them.
         backfill_message_ids = sorted(recent_message_ids - already_used_ids)
 
-        # Find which message ids we should mark as read.
-        # (We don't want too many unread messages.)
-        older_message_ids = set(backfill_message_ids[:-MAX_NUM_ONBOARDING_UNREAD_MESSAGES])
-
         # Create UserMessage rows for the backfill.
-        ums_to_create = []
         for message_id in backfill_message_ids:
-            um = UserMessage(user_profile=user_profile, message_id=message_id)
-            if message_id in older_message_ids:
-                um.flags = UserMessage.flags.read
+            um = UserMessage(
+                user_profile=user_profile,
+                message_id=message_id,
+                flags=UserMessage.flags.read | UserMessage.flags.historical,
+            )
             ums_to_create.append(um)
 
+    if len(ums_to_create) > 0:
         UserMessage.objects.bulk_create(ums_to_create)
```

Comment:
Would it make sense to sort these by `message_id` before creating them? Should have no functional impact, but just seems cleaner to do so. Equivalently, you could just have a single loop and set the flags based on whether the message_id is in `onboarding_message_ids`; that would probably be easier to read anyway.

Reply:
> Would it make sense to sort these by message_id before creating them?

Luckily, it is sorted in the current version because `onboarding_message_ids` is sorted due to `order_by("id")` + `backfill_message_ids = sorted(recent_message_ids - already_used_ids)`  + onboarding messages will always have lower ID value than other messages.

So, collectively it results in sorted by `message_id`.

> you could just have a single loop and set the flags based on whether the message_id is in onboarding_message_ids; that would probably be easier to read anyway.

It is possible that onboarding messages are not a part of the recent messages. Hence, two separate loops.

Reply:
or, maybe you're suggesting something like...
```
recent_and_onboarding_message_ids = sorted(recent_message_ids | onboarding_message_ids)

single loop and conditional flags
```

Reply:
The code for new algorithm aligns with it.